### PR TITLE
Add the restart and recompile commands.  Fixes #4

### LINKED ISSUE #4: Recompile and restart command
restricted to admins, add a command to recompile and restart (unless the compile failed)

### PR DESCRIPTION
**Pull request by SuperSpyTX**
See #4 

This works according to what was requested.  It exits with 14 (only) if the restart command is executed.  Compile errors go to both console & the log channel if necessary.  It does not automatically restart after the recompile.

You may need to refine the recompile.sh script accordingly, as this works when the current directory is in the GOPATH (go/src/github.com/riking/marvin).

I'm also not too sure about how to pass the flag for the exit code to occur (if the command was executed) without importing the module (see [here](https://github.com/riking/marvin/compare/master...SuperSpyTX:patch-restart-recompile?expand=1#diff-047912ae4e22f5323fe4d8a2a8004712)), so if that matters to you, I can try to find another way.

EDIT: Make sure you're redirecting stderr to stdout, or else Marvin won't have logs to read.  Eg: `2>&1`

### CODE REVIEW

File: modules/restart/restart.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+func (mod *RestartModule) RecompileCommand(t marvin.Team, args *marvin.CommandArguments) marvin.CommandResult {
+	if args.Source.AccessLevel() < marvin.AccessLevelAdmin {
+		return marvin.CmdFailuref(args, "This command is restricted to admins only.")
+	}
+
+	if RecompileInProgress {
+		return marvin.CmdFailuref(args, "There is an already existing recompile in progress.")
+	}
+
+	go mod.RecompileMarvin(t)
```
Comment (riking) [inline modules/restart/restart.go]: Just call the function and report back the results. "If you're waiting for the results of a computation, just call the function yourself." Commands don't have an inherent timeout.

File: modules/restart/restart.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	}
+
+	go mod.RestartMarvin(t)
+	return marvin.CmdSuccess(args, "I am restarting now, be back soon.")
+}
+
+func (mod *RestartModule) RecompileMarvin(t marvin.Team) {
+	RecompileInProgress = true
+	t.SendMessage(t.TeamConfig().LogChannel, "Recompiling Marvin....")
+	cmd := exec.Command("/bin/sh", os.Getenv("PWD") + "/recompile.sh")
```
Comment (riking) [inline modules/restart/restart.go]: The `recompile.sh` is located in the code, not the working directory (`~/marvin`).

Existing build script is named `build.sh`, we can use that. I just added `-e` to the script so it should give failures correctly now. So drop the recompile.sh

File: modules/restart/restart.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	"os"
+)
+
+func init() {
+	marvin.RegisterModule(NewRestartModule)
+}
+
+const Identifier = "restart"
+
+var RecompileInProgress = false
```
Comment (riking) [inline modules/restart/restart.go]: This variable causes data races; try compiling with the race detector `go install -race ...` or `go build -race ...`

File: modules/restart/restart.go (hunk 1 above)
Comment (riking) [reply modules/restart/restart.go]: No need to pass the team when it's a field of the module.

File: modules/restart/restart.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+	fmt.Printf("Compile Logs: \n%s", fmt.Sprintf("%s", stdout))
+	RecompileInProgress = false
+}
+
+func (mod *RestartModule) RestartMarvin(t marvin.Team) {
+	fmt.Printf("Restarting Marvin...")
+	RestartInProgress = true
+	t.SendMessage(t.TeamConfig().LogChannel, "Restarting Marvin...be back soon.")
+	syscall.Kill(syscall.Getpid(), syscall.SIGINT)
+}
```
Comment (riking) [inline modules/restart/restart.go]: Make sure to run `goimports`. The imports are also not sorted

File: modules/restart/restart.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+func (mod *RestartModule) Enable(team marvin.Team) {
+	cmd := team.RegisterCommandFunc("restart", mod.RestartCommand,
+		"`@marvin restart`"+
+			"This restarts the active Marvin instance.\n")
+	cmd2 := team.RegisterCommandFunc("recompile", mod.RecompileCommand,
+		"`@marvin recompile`"+
+			"This recompiles Marvin from git.\n")
+	team.RegisterCommand("restart", cmd)
```
Comment (riking) [inline modules/restart/restart.go]: This is for aliases, the command is already registered.

File: modules/restart/restart.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	team.RegisterCommand("recompile", cmd2)
+}
+
+func (mod *RestartModule) Disable(t marvin.Team) {
+}
+
+
+func (mod *RestartModule) RecompileCommand(t marvin.Team, args *marvin.CommandArguments) marvin.CommandResult {
+	if args.Source.AccessLevel() < marvin.AccessLevelAdmin {
```
Comment (riking) [inline modules/restart/restart.go]: `AccessLevelController`. Team owners should not be able to trigger restarts.

File: modules/restart/restart.go (hunk 2 above)
Comment (riking) [reply modules/restart/restart.go]: > EDIT: Make sure you're redirecting stderr to stdout, or else Marvin won't have logs to read. Eg: 2>&1

Call .CombinedOutput() https://godoc.org/os/exec#Cmd.CombinedOutput

Comment (SuperSpyTX) [reply modules/restart/restart.go]: One last question before I make my push: Where should I point this to? Build.sh or...?

File: modules/restart/restart.go (hunk 3 above)
Comment (SuperSpyTX) [reply modules/restart/restart.go]: I just realized this is the variable you were talking about.

File: modules/restart/restart.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+func (mod *RestartModule) RecompileCommand(t marvin.Team, args *marvin.CommandArguments) marvin.CommandResult {
+	if args.Source.AccessLevel() < marvin.AccessLevelController {
+		return marvin.CmdFailuref(args, "This command is restricted to controllers only.")
+	}
+
+	// This will check if it can take a buffer slot, if not, it means there's a recompile in progress.
+	// Otherwise it will recompile.
+	select {
+	case <-recompileSemaphore:
+		// defer reinserting the token until the recompile command is finished.
```
Comment (riking) [inline modules/restart/restart.go]: Dedent this code.

Make the body of the case a single `break`, and put the current body into the function body, after the select where all other paths lead to returns.

File: modules/restart/restart.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+		defer func() { recompileSemaphore <- struct{}{} }()
+		stdout, err := mod.Recompile()
+		if err != nil {
+			return marvin.CmdError(args, err, fmt.Sprintf("Failed to recompile: \n%s", stdout))
```
Comment (riking) [inline modules/restart/restart.go]: also send this to the log channel

Comment (SuperSpyTX) [reply modules/restart/restart.go]: I thought error does this by default.  I knew CmdFailure does.

Comment (riking) [reply modules/restart/restart.go]: oh right, it does I forgot

**Pull request merged**